File: web-static/js/blueprint.js
// Сборка чертежа Scrap Mechanic. Порт core/blueprint.py.
//
// Формат сверен с файлами, которые пишет сама игра:
//   {"bodies":[{"childs":[ ... ]}],"version":4}
//   child = {"bounds":{...},"color":"RRGGBB","pos":{...},
//            "shapeId":"<uuid>","xaxis":1,"zaxis":3}
// Цвет — 6 знаков в ВЕРХНЕМ регистре без решётки: сторонние конвертеры
// пишут "#FFFFFF", игра так не делает.
//
// Система координат игры — Z вверх, поэтому «картина» ложится в плоскость
// X-Z, а «мозаика на земле» — в X-Y.

export const VERTICAL = 'vertical';
export const HORIZONTAL = 'horizontal';

/** Ключ = упакованный цвет 0xRRGGBB, блок один на всю постройку. */
export function rgbResolver(shapeId) {
  return (key) => [((key >>> 0) & 0xffffff).toString(16).toUpperCase().padStart(6, '0'), shapeId];
}

/** Ключ = индекс в наборе материалов: у каждого свой цвет и свой блок. */
export function paletteResolver(palette, fallbackBlock) {
  return (key) => [
    (palette.paint[key] || '').toUpperCase(),
    palette.block[key] || fallbackBlock,
  ];
}

/**
 * Собрать текст blueprint.json.
 * Строки клеим вручную: на десятках тысяч деталей это на порядок быстрее и
 * экономнее по памяти, чем JSON.stringify по объектам.
 */
export function buildJson(rects, gridW, gridH, resolve, orientation = VERTICAL, center = true, depth = 1) {
  const offX = center ? -Math.floor(gridW / 2) : 0;
  depth = Math.max(1, depth | 0);

  const cache = new Map();
  const material = (key) => {
    let got = cache.get(key);
    if (!got) { got = resolve(key); cache.set(key, got); }
    return got;
  };

  const parts = new Array(rects.length);
  if (orientation === HORIZONTAL) {
    const offY = center ? -Math.floor(gridH / 2) : 0;
    for (let i = 0; i < rects.length; i++) {
      const [x, y, w, h, key] = rects[i];
      const [color, shape] = material(key);
      // строка 0 — верх картинки; сверху вниз = уменьшение Y
      parts[i] = `{"bounds":{"x":${w},"y":${h},"z":${depth}},"color":"${color}",`
        + `"pos":{"x":${x + offX},"y":${gridH - (y + h) + offY},"z":0},`
        + `"shapeId":"${shape}","xaxis":1,"zaxis":3}`;
    }
  } else {
    for (let i = 0; i < rects.length; i++) {
      const [x, y, w, h, key] = rects[i];
      const [color, shape] = material(key);
      // строка 0 — верх картинки; сверху вниз = уменьшение Z
      parts[i] = `{"bounds":{"x":${w},"y":${depth},"z":${h}},"color":"${color}",`
        + `"pos":{"x":${x + offX},"y":0,"z":${gridH - (y + h)}},`
        + `"shapeId":"${shape}","xaxis":1,"zaxis":3}`;
    }
  }

  return '{"bodies":[{"childs":[' + parts.join(',') + ']}],"version":4}';
}

/**
 * description.json ровно в том виде, в каком его пишет игра:
 * отступ в три пробела, порядок ключей, type = Blueprint.
 * Имя папки чертежа обязано совпадать с localId.
 */
export function descriptionJson(name, localId, note = '') {
  return JSON.stringify({
    description: note || 'Сделано в SM_Pixel — конвертер картинки в чертёж',
    localId,
    name,
    type: 'Blueprint',
    version: 0,
  }, null, 3);
}

/** Сколько деталей какого блока — для сводки в интерфейсе. */
export function usedBlocks(rects, resolve) {
  const counts = new Map();
  const cache = new Map();
  for (const [, , , , key] of rects) {
    let block = cache.get(key);
    if (block === undefined) { block = resolve(key)[1]; cache.set(key, block); }
    counts.set(block, (counts.get(block) || 0) + 1);
  }
  return counts;
}

/** UUID версии 4 — им называется папка чертежа. */
export function uuid4() {
  if (typeof crypto !== 'undefined' && crypto.randomUUID) return crypto.randomUUID();
  const b = new Uint8Array(16);
  (typeof crypto !== 'undefined' ? crypto : { getRandomValues: (a) => a.forEach((_, i) => { a[i] = (Math.random() * 256) | 0; }) })
    .getRandomValues(b);
  b[6] = (b[6] & 0x0f) | 0x40;
  b[8] = (b[8] & 0x3f) | 0x80;
  const hex = [...b].map((v) => v.toString(16).padStart(2, '0')).join('');
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

File: core/blueprint.py
"""Сборка чертежа Scrap Mechanic из прямоугольников.

Формат сверен с файлами, которые пишет сама игра (User/.../Blueprints и
Survival/LocalBlueprints):

blueprint.json
    {"bodies":[{"childs":[ ... ]}],"version":4}
    child = {"bounds":{"x":..,"y":..,"z":..},
             "color":"RRGGBB",              # ВЕРХНИЙ регистр, БЕЗ решётки
             "pos":{"x":..,"y":..,"z":..},  # левый-нижний-ближний угол блока
             "shapeId":"<uuid блока>",
             "xaxis":1,"zaxis":3}           # 1/3 = без поворота

description.json
    {"description":"...","localId":"<uuid>","name":"...",
     "type":"Blueprint","version":0}
    Имя папки чертежа обязано совпадать с localId.

Система координат игры — Z вверх. Значит:
    вертикально («картина»)   -> пиксели ложатся в плоскость X-Z
    горизонтально («на земле») -> пиксели ложатся в плоскость X-Y
"""

from __future__ import annotations

import json
import os
import shutil
import uuid as uuidlib

from .mesh import Rect

VERTICAL = "vertical"
HORIZONTAL = "horizontal"

_CHILD = (
    '{"bounds":{"x":%d,"y":%d,"z":%d},"color":"%s",'
    '"pos":{"x":%d,"y":%d,"z":%d},"shapeId":"%s","xaxis":1,"zaxis":3}'
)


def origin_of(grid_w: int, grid_h: int, orientation: str = VERTICAL, center: bool = True) -> tuple[int, int]:
    """Сдвиг картинки в локальных координатах чертежа: (по горизонтали, по вертикали).

    От него зависит, какая ячейка текстуры ляжет на какую клетку: развёртка
    в игре считается от локальных координат блока (проверено снимком).
    Клетка (x, y) картинки оказывается в локальных
        (x + ox,  grid_h - 1 - y + oz).
    Обе формулы обязаны совпадать с build_json ниже — иначе узор в
    предпросмотре разъедется с постройкой.
    """
    ox = -(grid_w // 2) if center else 0
    oz = (-(grid_h // 2) if center else 0) if orientation == HORIZONTAL else 0
    return ox, oz


def rgb_resolver(shape_id: str):
    """Ключ = упакованный цвет 0xRRGGBB, блок один на всю постройку."""
    return lambda key: ("%06X" % key, shape_id)


def palette_resolver(palette, fallback_block: str):
    """Ключ = индекс в наборе материалов: у каждого свой цвет и свой блок."""
    def resolve(key: int):
        paint = palette.paint[key]
        block = palette.block[key] or fallback_block
        return paint.upper(), block

    return resolve


def build_json(
    rects: list[Rect],
    grid_w: int,
    grid_h: int,
    resolve,
    orientation: str = VERTICAL,
    center: bool = True,
    depth: int = 1,
) -> str:
    """Собрать текст blueprint.json.

    resolve: ключ материала -> (цвет "RRGGBB", uuid блока). Результаты
    кэшируются: разных материалов всегда сильно меньше, чем деталей.

    Строки JSON клеим вручную: на десятках тысяч деталей это на порядок
    быстрее и экономнее по памяти, чем json.dumps со словарями.
    """
    if isinstance(resolve, str):                 # совместимость: передали uuid блока
        resolve = rgb_resolver(resolve)

    off_x = -(grid_w // 2) if center else 0
    parts: list[str] = []
    depth = max(1, int(depth))
    cache: dict[int, tuple[str, str]] = {}

    def material(key: int) -> tuple[str, str]:
        got = cache.get(key)
        if got is None:
            got = resolve(key)
            cache[key] = got
        return got

    if orientation == HORIZONTAL:
        off_y = -(grid_h // 2) if center else 0
        for x, y, w, h, key in rects:
            color, shape = material(key)
            # строка 0 — верх картинки; сверху вниз = уменьшение Y
            parts.append(_CHILD % (w, h, depth, color, x + off_x, grid_h - (y + h) + off_y, 0, shape))
    else:
        for x, y, w, h, key in rects:
            color, shape = material(key)
            # строка 0 — верх картинки; сверху вниз = уменьшение Z
            parts.append(_CHILD % (w, depth, h, color, x + off_x, 0, grid_h - (y + h), shape))

    return '{"bodies":[{"childs":[' + ",".join(parts) + ']}],"version":4}'


def used_blocks(rects: list[Rect], resolve) -> dict[str, int]:
    """Сколько деталей какого блока — для сводки в интерфейсе."""
    if isinstance(resolve, str):
        return {resolve: len(rects)}
    counts: dict[str, int] = {}
    cache: dict[int, str] = {}
    for _, _, _, _, key in rects:
        block = cache.get(key)
        if block is None:
            block = resolve(key)[1]
            cache[key] = block
        counts[block] = counts.get(block, 0) + 1
    return counts


def description_json(name: str, local_id: str, note: str = "") -> str:
    return json.dumps(
        {
            "description": note or "Сделано в SM_Pixel — конвертер картинки в чертёж",
            "localId": local_id,
            "name": name,
            "type": "Blueprint",
            "version": 0,
        },
        ensure_ascii=False,
        indent=3,
    )


def write_folder(
    target_dir: str,
    name: str,
    blueprint_text: str,
    icon: bytes | None = None,
    note: str = "",
    local_id: str | None = None,
    overwrite_same_name: bool = False,
) -> str:
    """Создать папку чертежа. Возвращает путь.

    Имя папки = localId, как это делает игра.
    """
    local_id = local_id or str(uuidlib.uuid4())
    path = os.path.join(target_dir, local_id)

    if overwrite_same_name:
        old = _find_by_name(target_dir, name)
        if old:
            shutil.rmtree(old, ignore_errors=True)

    os.makedirs(path, exist_ok=True)
    with open(os.path.join(path, "blueprint.json"), "w", encoding="utf-8") as fh:
        fh.write(blueprint_text)
    with open(os.path.join(path, "description.json"), "w", encoding="utf-8") as fh:
        fh.write(description_json(name, local_id, note))
    if icon:
        with open(os.path.join(path, "icon.png"), "wb") as fh:
            fh.write(icon)
    return path


def _find_by_name(target_dir: str, name: str) -> str | None:
    if not os.path.isdir(target_dir):
        return None
    for entry in os.listdir(target_dir):
        desc = os.path.join(target_dir, entry, "description.json")
        if not os.path.isfile(desc):
            continue
        try:
            with open(desc, encoding="utf-8") as fh:
                if json.load(fh).get("name") == name:
                    return os.path.join(target_dir, entry)
        except (OSError, ValueError):
            continue
    return None


def zip_bundle(items: list[dict], extras: dict[str, bytes | str] | None = None,
               lang: str = "ru") -> bytes:
    """Архив с одним или несколькими чертежами плюс произвольные файлы рядом.

    items: [{"name": ..., "text": ..., "icon": bytes|None, "note": ...}, ...]
    """
    import io
    import zipfile

    buf = io.BytesIO()
    folders = []
    with zipfile.ZipFile(buf, "w", zipfile.ZIP_DEFLATED) as z:
        for item in items:
            local_id = str(uuidlib.uuid4())
            folders.append(local_id)
            z.writestr(f"{local_id}/blueprint.json", item["text"])
            z.writestr(f"{local_id}/description.json", description_json(item["name"], local_id, item.get("note", "")))
            if item.get("icon"):
                z.writestr(f"{local_id}/icon.png", item["icon"])
        for path, data in (extras or {}).items():
            z.writestr(path, data)

        from . import i18n

        many = len(folders) > 1
        where = "%APPDATA%\\Axolot Games\\Scrap Mechanic\\User\\User_<SteamID>\\Blueprints\\"
        head = (i18n.t("doc.whereMany", lang, n=len(folders)) if many
                else i18n.t("doc.whereOne", lang))
        note = [head, where, i18n.t("doc.whereTail", lang)]
        if many:
            note += ["", i18n.t("doc.whereGuide", lang)]
        z.writestr(i18n.t("doc.whereName", lang), "\r\n".join(note) + "\r\n")

    return buf.getvalue()


def zip_bytes(name: str, blueprint_text: str, icon: bytes | None, note: str = "",
              lang: str = "ru") -> bytes:
    """Один чертёж архивом — чтобы перенести на другой компьютер."""
    return zip_bundle([{"name": name, "text": blueprint_text, "icon": icon, "note": note}], lang=lang)
